File: SegmentTree.java
import java.util.Arrays;

public class SegmentTree {
    int [] data;
    int [] orig;
    int n;
    public SegmentTree(int [] arr) {
        n = arr.length;
        orig = arr;
        data = new int[2 * n];

        System.arraycopy(arr, 0, data, n, n);

        /* skip 1st index */
        for (int i = n - 1; i > 0; --i)
            data[i] = Math.min(data[2 * i], data[2 * i + 1]);
    }

    public void update(int index, int v) {
        int i = index + n;
        data[i] = v;
        i = i/2;
        while (i > 0) {
            data[i] = Math.min(data[2 * i], data[2 * i + 1]);
            i = i/2;
        }
        printData();
    }

    private void printData() {
        for (int j = n ; j < 2 *n; j++)
            System.out.print(data[j] +", ");
        System.out.println();
    }

    /** assumes l and r are within bounds and r is exclusive. */
    public int findMin(int l, int r) {
        l += n;r += n;
        int min = Integer.MAX_VALUE;
        while (l <  r) {
            // if this is right node we include it and don't include the parent
            if (l % 2 == 1) {
                min = Math.min(min, data[l]);
                l++;
            }
            // Remember r is exclusive, this mean we are considering the left node
            // we include it and don't include the parent.
            if (r % 2 == 1) {
                r--;
                min = Math.min(min, data[r]);
            }
            l = l/2;
            r = r/2;
        }

        printData();
        return min;
    }

    public static void main(String[] args) {
        int[] arr = new int[]{1, 2, 3, 4, 5, 3, 1, 2, -1, 0};
        System.out.println(arr.length);
        SegmentTree  st = new SegmentTree(arr);

        System.out.println(Arrays.toString(arr));
    }
}

File: SegmentTree.scala
object SegmentTree extends App {

    class SegmentTree(_nums: Array[Int], empty: Int, combine: (Int, Int) => Int) {
        private val n = _nums.length
        private val data = Array.fill(n * 2)(0)
        Array.copy(_nums, 0, data, n, n)

        // build the tree in reverse order
        for (i <- (n- 1) to 1  by (-1)) data(i) = combine(data(2 * i), data(2 * i + 1))

        def update(index: Int, v: Int) {
            var i = n + index
            data(i) = v
            i = i / 2
            while (i > 0) {
                data(i) = combine(data(2 * i), data(2 * i + 1))
                i /= 2
            }
        }

        def rangeQuery(left: Int, right: Int): Int = {
            var l = left + n
            var r = right + n + 1
            var res = empty

            while (l < r) {
                // if this is right node we include it and don't include the parent
                if (l % 2 == 1) {
                  res = combine(res, data(l))
                  l += 1
                }
                // Remember r is exclusive, this mean we are considering the left node
                // we include it and don't include the parent.

                if (r % 2 == 1) {
                    r -= 1
                    res = combine(res, data(r))
                }

                l /= 2
                r /= 2
            }
            res
        }
    }


    val arr = Array(1, 3, 5)
    println(arr.mkString(","))

    def test1() = {
        val st = new SegmentTree(arr, 0, _ + _)
        println(st.rangeQuery(0, 2)) //
        st.update(1, 2)
        println(st.rangeQuery(0, 2))
    }

    def test2() = {
        val st = new SegmentTree(arr, Integer.MIN_VALUE, _ max _)
        println(st.rangeQuery(0, 2)) //
        println(st.rangeQuery(1, 2))
        st.update(1, 2)
        println(st.rangeQuery(1, 2))
    }

    test1()
    test2()
}